[benchmark][LoadGenerator] Process both read/write load requests

littlepretty opened the pull request:
## Summary:
This commit stack introduces the new features to make Benchmarker to be able to handle generic LoadRequest. It spawns a thread to handle read requests' responses.

## Test Plan:
Added a smoke test.

Also tested by adding read requests in both ring and pairwise generators, use a histogram counter to record the size of the response items.
```
$ ./target/release/ruben -f faucet_for_ruben -s temp_config -g 100 -n 32 -c 4 -e 6 \
-m localhost:45345 -t pairwise -q 3 -k 40
$ ./target/release/ruben -f faucet_for_ruben -s temp_config -g 100 -n 32 -c 4 -e 6 \
-m localhost:45345 -t ring -q 3 -k 40
```
![image](https://user-images.githubusercontent.com/1838298/62438519-a3f0ec80-b6fc-11e9-8b1d-cf91663e9ad4.png)

0-o-0 commented on benchmark/src/grpc_helpers.rs:
```diff
@@ -92,25 +92,65 @@ fn check_ac_response(resp: &ProtoSubmitTransactionResponse) -> bool {
 /// Ignore but count both gRPC-failed submissions and AC-rejected TXNs.
 pub fn submit_and_wait_txn_requests(
```
it's not only for txn_requests now, let's update the function name

fbandersnatch replied:
There are a few methods that refer to txn_requset / txn_load -- these should be updated, since reads are not really transaction requests.

littlepretty replied:
Please see my general response about renaming.

0-o-0 commented on benchmark/src/grpc_helpers.rs:
```diff
@@ -92,25 +92,65 @@ fn check_ac_response(resp: &ProtoSubmitTransactionResponse) -> bool {
 pub fn submit_and_wait_txn_requests(
     client: &AdmissionControlClient,
-    txn_requests: Vec<SubmitTransactionRequest>,
+    txn_requests: Vec<LoadRequest>,
```
`txn_requests` -> `load_requests`

littlepretty replied:
See my general response about renaming.

0-o-0 commented on benchmark/src/grpc_helpers.rs:
```diff
@@ -92,25 +92,65 @@ fn check_ac_response(resp: &ProtoSubmitTransactionResponse) -> bool {
+    }
+
+    // Process read load's responses. Main thread won't join, hence not blocked by this thread.
+    let read_stream = stream::futures_unordered(read_futures);
```
as the function gets longer, I think it's better to extract `read_futures` and `write_futures` processing to dedicated functions.

littlepretty replied:
Sure. Extracted to `wait_read/write_futures()`.

fbandersnatch commented:
From offline discussion:
I suggest separating the added functionality (i.e., full support of `LoadRequest`, without conversion to `SubmitTransactionRequest`) from the example use that introduces another argument and changes the existing transaction generators.

littlepretty commented:
About renaming **TXNs** => **load requests**:
* I renamed `TXNs` to `load requests` almost everywhere, since load covers both read load and TXN loads, including both code and comments.
* The module is also renamed to `load_generator.rs`.
* Except for the case that the load request can only be TXNs. For example, RingTransferTxnGenerator/PairwiseTransferTxnGenerator are generating only TXNs, so `TXN` still suite them. Another examples include `wait_txns_committed` and counters for TXNs, since they are TXN specific.
* In `ruben_opt.rs`, the argument is still `TransactionPattern` because load are only from TXN generator for now, but it will be changed to `LoadPattern` sooner or later.


Remove the `-q` parameter and its corresponding feature, leave existing txn generator unchanged.

littlepretty commented:
Rebase to upstream/master.

littlepretty commented on benchmark/src/ruben_opt.rs:
```diff
@@ -73,10 +73,10 @@ pub struct Opt {
     #[structopt(short = "k", long = "submit_rate")]
     pub submit_rate: Option<u64>,
-    /// Choices of how to generate TXNs/load.
+    /// Choices of how to generate TXNs. TODO: extend to load request pattern.
```
Currently we only have Ring/Pairwise TXN generators. In future, when we have read request generators, we should rename `TransactionPattern` to `RequestPattern` and `txn_pattern` to `request_pattern`.

littlepretty commented on benchmark/src/bin_utils.rs:
```diff
@@ -64,32 +64,32 @@ pub fn try_start_metrics_server(args: &Opt) {
     bm.mint_accounts(&mint_txns, faucet_account);
 
-    // Submit TXN load and measure throughput.
+    // Submit pure TXNs and measure throughput.
```
Yes, they mean the same thing. Perhaps I should just use the name transactions, to not confuse people?

littlepretty commented:
> I don't want us to invent new terminology, and "load request" shouldn't become a thing (since we do not have requests that are free and not generating any load).
> 

OK. I have to admit I am very bad at naming things. 
Just to confirm, @fbandersnatch 
1. we should change `LoadGenerator` and `load_generator` into `RequestGenerator` and `request_generator.rs`? 
2. and `struct LoadRequest` should be changed to `struct Request`?

> It seems like "load/loads" terminology used here is interchangeable with "requests". Can we simply use requests everywhere, so no one needs to learn new terms?

I will sweep the code to replace `load` with `request` in comments.
Also I will avoid using different words when referring to TXN.

fbandersnatch commented:
LoadGenerator is a fine name. But as part of your name changes in the code + comments you passed through and replaced the word "transaction" or "request" with the word "load", and I don't think that it was a good change -- it's a new non standard name, sometimes used in combination with other names to make it just confusing.

littlepretty commented on benchmark/src/bin_utils.rs:
```diff
@@ -64,32 +64,32 @@ pub fn try_start_metrics_server(args: &Opt) {
 }
 
-/// Play given TXN pattern with Benchmarker for several epochs and measure burst throughput,
+/// Play given TXN loads with Benchmarker for several epochs and measure burst throughput,
```
Changed to "Play given TXNs with Benchmarker..."

littlepretty commented on benchmark/src/grpc_helpers.rs:
```diff
@@ -86,6 +88,57 @@ fn check_ac_response(resp: &ProtoSubmitTransactionResponse) -> bool {
 }
 
+/// Process read load's responses in a separate thread.
+fn wait_read_requests(read_futures: Vec<ClientUnaryReceiver<UpdateToLatestLedgerResponse>>) {
```
Sure.

littlepretty commented on benchmark/src/grpc_helpers.rs:
```diff
@@ -86,6 +88,57 @@ fn check_ac_response(resp: &ProtoSubmitTransactionResponse) -> bool {
+                        "Receive {} bytes of UpdateToLatestLedgerResponse",
+                        resp_size
+                    );
+                    for item in proto_resp.get_response_items().iter() {
```
Not too much difference. It maybe useful if we want to check the individual response item. But here it doesn't and we should remove it.

littlepretty commented:
> LoadGenerator is a fine name. But as part of your name changes in the code + comments you passed through and replaced the word "transaction" or "request" with the word "load", and I don't think that it was a good change -- it's a new non standard name, sometimes used in combination with other names to make it just confusing.

Ok. I removed "load(s)" from comments/code, except for "load_generator.rs" and "LoadGenerator".

Also rebased to upstream/master.

fbandersnatch commented on benchmark/src/bin/ruben.rs:
```diff
@@ -67,8 +68,7 @@ mod tests {
         fn test_benchmarker_counters() {
-            let (faucet_account_keypair, faucet_key_file_path, _temp_dir) =
-                generate_keypair::load_faucet_key_or_create_default(None);
+            let (faucet_account_keypair, faucet_key_file_path, _temp_dir) = generate_keypair::load_faucet_key_or_create_default(None);
```
Does it let you do such long lines without complaining?

littlepretty replied:
It seems the checks are only checking the squashed commit, I mean the entire PR.
